esp/logging/loggingagent/cassandraloggingagent/cassandralogagent.cpp

```diff
@@ -0,0 +1,621 @@
+{
+    unsigned retry = 1;
+    while (1)
+    {
+        try
+        {
+            const char* appName = req.getApplication();
+            if (!appName || !*appName)
+                appName = defaultTransactionApp.get();
+
+            StringBuffer logSeed;
+            queryTransactionSeed(appName, logSeed);
+
+            if (!logSeed.length())
+                throw MakeStringException(EspLoggingErrors::GetTransactionSeedFailed, "Failed to get TransactionSeed");
+
+            resp.setSeedId(logSeed.str());
+            resp.setStatusCode(0);
+            return true;
+        }
+        catch (IException* e)
+        {
+            StringBuffer errorStr, errorMessage;
+            errorMessage.append("Failed to get TransactionSeed: error code ").append(e->errorCode()).append(", error message ").append(e->errorMessage(errorStr));
+            ERRLOG("%s -- try %d", errorMessage.str(), retry);
+            e->Release();
+            if (retry < maxTriesGTS)
+            {
+                Sleep(retry*3000);
+                retry++;
+            }
+            else
+            {
+                resp.setStatusCode(-1);
+                resp.setStatusMessage(errorMessage.str());
+                break;
+            }
+        }
+    }
+    return false;
+}
+
+bool CCassandraLogAgent::updateLog(IEspUpdateLogRequestWrap& req, IEspUpdateLogResponse& resp)
+{
+    try
+    {
+        StringBuffer requestBuf = req.getUpdateLogRequest();
+        if (requestBuf.isEmpty())
+            throw MakeStringException(EspLoggingErrors::UpdateLogFailed, "Failed to read log request.");
+
+        StringBuffer logDB, logSource;
+        requestBuf.insert(0, "<LogRequest>");
+        requestBuf.append("</LogRequest>");
+        Owned<IPropertyTree> logRequestTree = createPTreeFromXMLString(requestBuf.length(), requestBuf.str());
+        if (!logRequestTree)
+            throw MakeStringException(EspLoggingErrors::UpdateLogFailed, "Failed to read log request.");
+
+        CLogGroup* logGroup = checkLogSource(logRequestTree, logSource, logDB);
+        if (!logGroup)
+            throw MakeStringException(EspLoggingErrors::UpdateLogFailed, "Log Group %s undefined.", logSource.str());
+
+        StringBuffer logID;
+        getLoggingTransactionID(logID);
+
+        CIArrayOf<CLogTable>& logTables = logGroup->getLogTables();
+        setKeySpace(logDB.str());
+        ForEachItemIn(i, logTables)
+        {
+            CLogTable& table = logTables.item(i);
+
+            StringBuffer cqlStatement;
+            if(!buildUpdateLogStatement(logRequestTree, logDB.str(), table, logID, cqlStatement))
+                throw MakeStringException(EspLoggingErrors::UpdateLogFailed, "Failed in creating SQL statement.");
+
+            if (getEspLogLevel() >= LogMax)
+                DBGLOG("CQL: %s\n", cqlStatement.str());
+
+            cassSession->connect();
```

Comment:
if possible we should cache/pool. Maybe cassandra connections aren't as expensive as other datastores?